partner_tag_smart_assignation/tests/test_smart_tagger.py

```diff
@@ -0,0 +1,161 @@
+                "author_id": False,
+                "department_ids": False,
+                "description": "",
+                "valid_until": fields.Date.today(),
+                "tag_filter_partner_field": "partner_id",
+                "tag_filter_condition_id": self.create_condition().id,
+            }
+        )
+
+    def test_create_tag(self):
+        """
+        Tag all partners which have the letter 'o' in the name.
+        """
+        smart_tag = self.create_tag()
+        for partner in self.partners.filtered(lambda t: "o" in t.name):
+            self.assertTrue(partner in smart_tag.partner_ids)
+
+        for partner in self.partners.filtered(lambda t: "o" not in t.name):
+            self.assertFalse(partner in smart_tag.partner_ids)
+
+        for partner in smart_tag.partner_ids:
+            self.assertTrue("o" in partner.name)
+
+    def test_modify_partner(self):
+        """
+        Tag partners with the letter 'o' in the name.
+        Then edit some name and check whether the smart tag updates
+        """
+
+        smart_tag = self.create_tag()
+
+        # update some first names
+        michael = self.michel_fletcher
+        michael.update({"name": "Michel Angelo"})
+
+        # Simulate a cron trigger
+        self.env["res.partner.category"].update_all_smart_tags()
+
+        # verify that the updated tag contains the partner 'Michel Angelo'
+        self.assertTrue(michael in smart_tag.partner_ids)
+
+        for partner in smart_tag.partner_ids:
+            self.assertTrue("o" in partner.name)
+
+    def test_smart_tag_sql(self):
+        """Test query SQL for smart tags"""
+        smart_tag = self.env["res.partner.category"].create(
+            {
+                "name": "Test Smart Tag SQL",
+                "active": True,
+                "smart": True,
+                "parent_id": False,
+                "tag_filter_sql_query": """
+                SELECT id
+                FROM res_partner
+                Where name like '%o%'
+            """,
+            }
+        )
+        # update some first names
+        michael = self.michel_fletcher
+        michael.write({"name": "Michel Angelo"})
+
+        # Trigger tag update
+        smart_tag.update_partner_tags()
+
+        # verify that the updated tag contains the partner 'Michel Angelo'
+        self.assertTrue(michael in smart_tag.partner_ids)
+
+        for partner in smart_tag.partner_ids:
+            self.assertTrue("o" in partner.name)
+
+    def test_check_validity_dates(self):
+        """
+        Test if the valid_until functionality works correctly
+        """
+
+        # Create a new tag with a 'valid_until' date set to yesterday
+        yesterday = fields.Date.to_string(fields.Date.today() - timedelta(days=1))
+
+        expired_tag = self.create_tag()
+        expired_tag.update({"valid_until": yesterday})
+        expired_tag.update_partner_tags()
+
+        # Reload the tags from the database
```

Comment:
This is not needed. Actually updating the database through write, will update the database (and the cache). invalidate_cache can be useful after changing stuff through SQL.